File: javascript/0103-binary-tree-zigzag-level-order-traversal.js
/**
 * BFS
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = (root) => {
    const isEdgeBase = (root === null);
    if (isEdgeBase) return [];
    
    return search(root);/* Time O(N) | Space O(N) */
};

var search = (root, isZigZag = true, order = []) => {
    const queue = new Queue([ root ]);
    
    while (!queue.isEmpty()) {        /* Time O(N) */
        const levels = [];

        bfs(queue, isZigZag, levels); /* Time O(WIDTH) | Space O(WIDTH) */
        order.push(levels);           /*                 Space O(N) */
        isZigZag = !isZigZag;
    }
    
    return order;
}

const bfs = (queue, isZigZag, levels) => {
    for (let level = queue.size(); (0 < level); level--) {/* Time O(WIDTH) */
        const { left, val, right } = queue.dequeue();
    
        if (left) queue.enqueue(left);  /* Space O(WIDTH) */
        if (right) queue.enqueue(right);/* Space O(WIDTH) */
        
        levels.push(val);               /* Space O(N) */
    }

    if (!isZigZag) levels.reverse();
}

File: java/0103-binary-tree-zigzag-level-order-traversal.java
class Solution {
    boolean reverse = false; // flag to detrmine the direction left or right
    List<List<Integer>> sol = new ArrayList<List<Integer>>();
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if(root == null){return sol;}
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while( !queue.isEmpty()){  
            List<Integer> temp = new ArrayList<Integer>();
            int size =queue.size();
            for(int i=0; i < size; i++){
                TreeNode node = queue.poll();
                temp.add(node.val);
                if(node.left != null){
                    queue.add(node.left);
                }
                if(node.right != null){
                    queue.add(node.right);
                }
            }
            if(reverse){
                Collections.reverse(temp);
            }
            reverse = !reverse;
            sol.add(temp);
        }
        return sol;
    }

}